### ISSUE TITLE
kraken orderbook updates are different than bitfinex

### ISSUE THREAD

**Issue by wisepotato** [altangent/ccxws#76]
**Exchange**
kraken

**Subscription type**
L2updates and L2snapshots

**Describe the bug**
bitfinex/kraken have different updates. When using bitfinex and L2updates and L2snapshots, i get a working, valid orderbook. When I use kraken in the same system, it doesnt work/update. What is going on there? Do i need to incorporate trades?

** steps to reproduce**
cant really reproduce as its production code, but is there documentation missing? I'm leaning towards implementing the websockets of kraken myself as binance and bitfinex both work but kraken is just not working properly.

Is there something i'm missing?

**Comment by bmancini55** [comment]
You do no need to incorporate trades to get orderbooks to work. Most likely you are not sending a valid market.  

```

```

**Comment by wisepotato** [comment]
Man, first of all; thanks for the extensive reply. I'm checking my ids, and its `XXBTZUSD` for kraken and `BTCUSD` for bitfinex,  does the base/quote matter btw? Its just a marker for internal use right? I do exactly what you do, but i also subscribe for snapshots ofcourse `subscribeLevel2Snapshots`

**Comment by wisepotato** [comment]
so to be clear, i am getting information from kraken, i just think its malformed. I use an RBTree to update my orderbook and its just not updating correctly for kraken. Do i need to take trades into account for kraken? Maybe i'm doing wrong stupidly. I even thought about implementing the webhooks directly but i'd rather use ccxws :)

**Comment by bmancini55** [comment]
Glad to investigate with you. The markets sound correct and since you're getting data, my guess was not correct haha!  As you mentioned, the `base`/`quote` market properties are just markers that get passed with the result object.  So that shouldn't matter.

You don't need trades since the raw order book updates stream uses the absolute volume at a price level and not the volume delta. The volume will be also be 0 when a price point should be removed.  Therefore, the update stream will take into account any trade matches.

The interesting thing with Kraken is that the raw update stream includes a nanosecond timestamp at each price level. This value indicates the timestamp of change for each price point.  

I'm not sure if it matters that CCXWS prunes this data.  Instead, we just find the maximal timestamp and assigns that to the `timestampMs` property of the overall update. The assumption is that the to maintain an order book only requires a check to make sure the OB's timestamp is less than the update payload's timestamp.  

It might be an interesting exercise to try it with a raw stream and do the timestamp checks and see if the OB is more consistent.

The Kraken documentation mentions "republished" values: https://www.kraken.com/en-us/features/websocket-api#message-book.  I'm not exactly sure what the purpose of that is, but if for some reason it broadcast out of date values, it is possible that without the per price timestamps, old entries could be broadcast.

Lastly, I know one issue that we've had (I don't recall if it was with Kraken) is that dropped messages from the exchange can cause you to miss deletes in a fast moving market.  The end result is bid/ask overlap that won't clear.  We have logic to self heal the order book by clear overlaps.

**Comment by wisepotato** [comment]
> You don't need trades since the raw order book updates stream uses the absolute volume at a price level and not the volume delta. The volume will be also be 0 when a price point should be removed. Therefore, the update stream will take into account any trade matches.

Thanks for this, maybe this should be more obvious in the readme? I had trouble knowing whether or not to include the trades, but this helps.

What I did notice is that my orderbook is more up to date when I get rid of logging (therefore it becomes quicker). But then still I had some issues after a while. This was on a mobile I7 within the last year so I didnt think it would matter.

> The end result is bid/ask overlap that won't clear. We have logic to self heal the order book by clear overlaps.

What do you mean by this last part? Is that  a part of logic I have to implement or is it done automagically?

**Comment by bmancini55** [comment]
Sure thing, I can add a section on building order books to the README.  Each one is a bit different depending on the data that is emitted by the exchange.  So even with CCXWS it still requires some custom code to get things working consistently.

You will have to implement recovery logic on your own. 

For exchanges that do broadcast a sequenceId you can check each update at the correct sequence. If not, you have lost a message and need to restart the feed to recover the order book. 

For exchanges that don't implement a sequence you cannot determine if you have dropped a message immediately. Your order book will eventually get out of sync and you will end up with bid/ask overlap.

For example, initial state:
```
      11002
      11001 
Asks: 11000
Bids: 10999
      10998
      10997 
````

Update 1 - missed
```
Asks: 11001 > delete
      11000 > delete
 
```

Update 2 - received
```
Bids: 11001 > add
```

Because update 1 was missed there are ask values that should have been removed (11000 and 11001).  When update 2 is received these ask values overlap with the bid value (11001).  So you can detect when the order book has gone out of sync by checking for best bid and best ask overlap. 

One recovery mechanism is to prune the side that is overlapping. In this example, since a bid update of 11001 is received in update 2, we know that any ask value <= 11001 should be removed and we can clear those values. This doesn't guarantee that the rest of the order book is consistent (there could have been other changes to the order book that we missed). 

Alternatively, you could perform a reconnect and re-sync the order book from the snapshot.

**Comment by wisepotato** [comment]
thank you very much for your reply! I didnt know this, this would probably be the cause of my woes, it is a shame that without a sequenceId we dont have a way to check for missed messages.

Is this a speed issue? Because I did notice that the less console.logs i did the faster my updates became

**Comment by bmancini55** [comment]
It could be a speed issue, I haven't done much research on how client backpressure impacts various exchanges. My speculation would be that the server would start dropping messages when it receives back pressure.

**Comment by bmancini55** [comment]
Closing issue.  Adding new section to README in #92

**Comment by wisepotato** [comment]
To be fair, ive abandoned this package (sadly) because the orderbook wouldnt be correct in the most simplest built orderbook. I could not for the life of me get the orderbook to be correct, so im.resorting to implementing thr api directly

**Comment by 49e94b8f256530dc0d41f740dfe8a4c1** [comment]
@wisepotato I thought I was the only one. My orderbook also always went out of sync, when a spike in orders on the exchange cleared most of the scope(bots tripping over each other) or a large buy that eats up most orders. It seems the client just dropped messages since it could not handle the spike. But the orderbook stayed in sync for days for low volume markets

**Comment by tomekit** [comment]
Hi @wisepotato
Just wondering whether you've managed to build reliable orderbook on top of Kraken WS feed?
Came across this library, but wrongly assumed it already manages the orderbook state. 

I am trying to implement the orderbook, need to admit that most of the exchanges I've got covered with success, but I am struggling to keep order book in sync with Kraken.

I would appreciate if you could share some experience. 
Is `timestamp` needed to keep the order/sequence? What's the `republish`? Does cross bid/ask happen to you? If so, what's your self-heal mechanism?
Thanks

**Comment by tomekit** [comment]
Hi @wisepotato
Just wondering whether you've managed to build reliable orderbook on top of Kraken WS feed?
Came across this library, but wrongly assumed it already manages the orderbook state.

I am trying to implement the orderbook, need to admit that most of the exchanges I've got covered with success, but I am struggling to keep order book in sync with Kraken.

I would appreciate if you could share some experience.
Is timestamp needed to keep the order/sequence? What's the republish? Does cross bid/ask happen to you? If so, what's your self-heal mechanism?

**Comment by wisepotato** [comment]
yeah i've started doing it myself as ccxws is awesome but it makes a lot of assumptions about knowledge, and it was hard to get into the raw details of hjow ccxws does stuff.

I would suggest doing it yourself. Its probably also faster, and less generic

**Comment by Rafael-Ayala** [comment]
Hi, I am currently grabbing data from the Kraken websocket API and have found issues potentially related to what was discussed previously in this thread. Basically, I have managed to correctly receive and parse the streamed data regarding order books. I have verified that I am receiving and parsing correctly by calculating at each update the checksum for the updated book as described [here](https://docs.kraken.com/websockets/#book-checksum) and comparing with the broadcasted checksum. I get a matching checksum almost always, but on very rare occasions I get a different checksum. It is quite problematic, because after one of these rare events, the checksums don't match again until a fresh book is requested from 0 again, meaning that somehow I process something differently. This seems to be independent of the exchange pair(s). I am currently still troubleshooting it and trying to find out what triggers them, but a very interesting observation I have recently made is that the checksum mismatches seem to always happen when either the most recent update message, or the one before that, contains one of those "republished" updates mentioned by @bmancini55 . I am wondering if anyone figured out what exactly they are, or how they should be dealt with? I thought they just had to be inserted in the order book as any other update (i.e., as normal, non "republished" updates), but apparently I'm missing something... Thanks!

**Comment by simbesh** [comment]
Just in case someone stumbles upon this thread like I did the answer is here:

https://blog.kraken.com/post/2341/new-version-of-kraken-websockets-public-api-goes-live-may-30/#:~:text=The%20order%20book%20feed%20includes%20an%20additional%20field%20when%20price%20levels%20are%20pulled%20into%20scope%20by%20another%20price%20level%20being%20deleted%20(also%20known%20as%20a%20border%20republish)

tldr: when a price deletion causes a price that was outside your depth to fall into scope, kraken will send you the "missing" prices
